### DISCUSSION TITLE
Collections without business ID

### DISCUSSION THREAD

**Discussion by JDMKSZ** [belgif/rest-guide#187] [General]
The Belgif REST guide has a rule on collection representation (https://www.belgif.be/specification/rest/api-guide/#rule-col-repres[col-repres]). This rule says that each object in an item list MUST have a link to the individual resource, and SHOULD have a business identifier. Note that the main reason for the requirement is without it, no individual resource of the collection can be updated.

However, sometimes it doesn't make sense to show an individual item from a collection, or it has no URL because it isn't a separate resource and/or has no business identifier. We distinguish several use cases here, with for each them some examples and a proposed approach.

###  Relation

| Entity 1 | Entity 2 | Relation attributes |
| -------- | ------- | ------- |
|Blog article|Author|Author type (enum or refData resource, e.g. author/co-author/editor)|
|Ticket|User|Involved as (enum or refData, e.g. creator/updater/watcher)|
|Prescription|Healthcare provider|Issuer/Responsible (enum or refData)|

The second entity has a business identifier, but the relation itself hasn't, it isn't even a resource. So if the item contains a link to the second entity/resource, this won't contain the relation attributes any more but not to the "relation" resource itself properties itself. Note that using the business identifier of the second entity as subresource identifier isn't an option because it can appear multiple times.

#### Proposed approach:

Either
`/article/{articleId}`
returning
```
{
  "title": "...",
  "body": "...",
  "author": 15,
  "coAuthors": [22, 105, 31]
}
```

so the relationship properties were changed into main properties with that name, this would only work for a single relationship property like the relationship type. Otherwise, check if the next chapters aren't better suited.

### History

History can be considered as a specific case of a relation between two resources, or a resource and a subresource, with the period as an attribute of the relationship, e.g.

| Entity 1 | Entity 2 | Relation attributes |
| -------- | ------- | ------- |
|Company|Employee|Period, role, wage, ...|
|Person or company|Address|Address type, start date, end date|

Note that the same entity (e.g. same address) can appear multiple times in the history (but with a different period).

#### Proposed approach (if you cannot have an ID per record in the history):

`/company/{companyId}/employee/{employeeId}`
returns the current situation:
```
{
  "employeeId": 123456,
  "personId": 11111111111, # This might refer to /person/{personId}
  "role": "CEO",
  "startDate": "2024-06-19",
  ...
}
```
and
`/company/{companyId}/employee/{employeeId}/history`
returns all the situations for this employee:
```
[
  {
    "employeeId": 123456,
    "personId": 11111111111, # This might refer to /person/{personId}
    "role": "CEO",
    "startDate": "2024-06-19",
    ...
  },
  {
    "employeeId": 123456,
    "personId": 11111111111, # This might refer to /person/{personId}
    "role": "CFO",
    "startDate": "2000-01-01",
    "endDate": "2024-06-18"
  }
]
```

The history is not considered to be a normal collection here. To update the history, you need to a `PUT /company/{companyId}/employee/{employeeId}/history` of the entire new history. This approach doesn't allow individual history records to be updated (or inserted/removed).

### View / aggregation

Composed views and aggregations are combinations of multiple resources into a single collection view, for which the individual resources may have identifiers (or not, e.g. the examples below where the data is coming from the National Register), but the combined resources haven't.

Examples:

* The household history "view" the CBSS offers based on the National Register data
* The address history the CBSS currently offers, based on multiple "IT's" of the national register, combined with the address in the CBSS registries.
* Aggregated data per quarter of employment (DmfA/Dimona)
* Career/employment summary + totals aggregated per year

#### Proposed approach:

As the elements in this collection never need to be updated, it's OK to have no individual links or business ID.

### Next steps

1. Does this make sense to all of you?
2. Should we add these solutions to the REST guide? Or just as examples?
3. In [[col-repres]](https://www.belgif.be/specification/rest/api-guide/#rule-col-repres), are we going to change the requirement "In these objects, a href property MUST be present with a link to the resource." to "SHOULD"?

**Comment by pvdbosch** [comment]
I think its mostly a matter of definition and modeling: the guide defines a collection resource as "a server-managed list of document resources", while a document resource "is a singular concept that is referring to a business entity or object instance" and always has a primary key (which may be composite, and include its parent resource's key). 

Most above cases are (document) resources that have lists in their representation, but their items aren't documents so the [col-repres] rule doesn't apply to them.
Defining document resources is useful for consistent interpretation and designing the lifecycle of data (e.g. CRUD operations, reuse of data), but can't be shoehorned onto everything for a variety of possible reasons (technical, legacy, business, ...).

The above case of "history" (where there's no primary per history entry), is modeled as a [singleton document resource](https://www.belgif.be/specification/rest/api-guide/#document-singleton) then, rather as a collection.

**Comment by JDMKSZ** [comment]
So in the above cases where we have a singleton document resource which contains a list, what should be the name of the collection? Should it always be a custom name, to avoid confusion (e.g. `person/123/addressHistory` returns `{addresses: [ ...]}` ? Or can we use `items`? In other words, is it better to align as much to the naming (`items`) or should we avoid this (and maybe explicitly forbid it) if it's not a collection of documents, to avoid confusion?

**Reply by pvdbosch** [reply]
There's no rule on this, something can be said for either option. 
If there's a single list in the resource of which the content is clear from the URL, `items` could be used and optionally also pagination properties. If there are multiple lists, or the list content is not clear from the URL, a more specific name than `items` could be needed.

**Reply by jpraet** [reply]
https://www.belgif.be/specification/rest/api-guide/#rule-col-repres does say that "a href property MUST be present with a link to the resource" for the objects in the `items` array. If such rule ever gets enforced by the https://github.com/belgif/rest-guide-validator, it would trigger a violation if we use an `items` array in other collection-like resources without those hrefs. Aside from that, I would prefer to use `items`.
